src/sagemaker_server.cc

```diff
@@ -138,4 +283,285 @@ SagemakerAPIServer::Create(
+
+  if (unload_err != nullptr) {
+    EVBufferAddErrorJson(req->buffer_out, unload_err);
+    evhtp_send_reply(req, EVHTP_RES_BADREQ);
+    TRITONSERVER_ErrorDelete(unload_err);
+  }
+
+  sagemaker_models_list.erase(model_name);
+}
+
+void
+SagemakerAPIServer::SageMakerMMEGetModel(
+    evhtp_request_t* req, const char* model_name)
+{
+  if (sagemaker_models_list.find(model_name) == sagemaker_models_list.end()) {
+    evhtp_send_reply(req, EVHTP_RES_NOTFOUND); /* 404*/
+    return;
+  }
+
+  triton::common::TritonJson::Value sagemaker_get_json(
+      triton::common::TritonJson::ValueType::OBJECT);
+
+  sagemaker_get_json.AddString("modelName", model_name);
+  sagemaker_get_json.AddString(
+      "modelUrl", sagemaker_models_list.at(model_name));
+
+  const char* buffer;
+  size_t byte_size;
+
+  triton::common::TritonJson::WriteBuffer json_buffer_;
+  json_buffer_.Clear();
+  sagemaker_get_json.Write(&json_buffer_);
+
+  byte_size = json_buffer_.Size();
+  buffer = json_buffer_.Base();
+
+  evbuffer_add(req->buffer_out, buffer, byte_size);
+  evhtp_send_reply(req, EVHTP_RES_OK);
+}
+
+void
+SagemakerAPIServer::SageMakerMMEListModel(evhtp_request_t* req)
+{
+  triton::common::TritonJson::Value sagemaker_list_json(
+      triton::common::TritonJson::ValueType::OBJECT);
+
+  triton::common::TritonJson::Value models_array(
+      sagemaker_list_json, triton::common::TritonJson::ValueType::ARRAY);
+
+  for (auto it = sagemaker_models_list.begin();
+       it != sagemaker_models_list.end(); it++) {
+    triton::common::TritonJson::Value model_url_pair(
+        models_array, triton::common::TritonJson::ValueType::OBJECT);
+
+    bool ready = false;
+    TRITONSERVER_ServerModelIsReady(
+        server_.get(), it->first.c_str(), 1, &ready);
+
+    /* Add to return list only if model is ready to be served */
+    if (ready) {
+      model_url_pair.AddString("modelName", it->first);
+      model_url_pair.AddString("modelUrl", it->second);
+    }
+
+    models_array.Append(std::move(model_url_pair));
+  }
+
+  sagemaker_list_json.Add("models", std::move(models_array));
+
+  const char* buffer;
+  size_t byte_size;
+
+  triton::common::TritonJson::WriteBuffer json_buffer_;
+  json_buffer_.Clear();
+  sagemaker_list_json.Write(&json_buffer_);
+
+  byte_size = json_buffer_.Size();
+  buffer = json_buffer_.Base();
+
+  evbuffer_add(req->buffer_out, buffer, byte_size);
+  evhtp_send_reply(req, EVHTP_RES_OK);
+}
+
+void
+SagemakerAPIServer::SageMakerMMELoadModel(
+    evhtp_request_t* req,
+    const std::unordered_map<std::string, std::string> parse_map)
+{
+  std::string repo_path = parse_map.at("url");
+  std::string model_name = parse_map.at("model_name");
+
+  /* Error out if there's more than one subdir/version within
+   * supplied model repo, as ensemble in MME is not (currently)
+   * supported
+   */
+  DIR* dir;
+  struct dirent* ent;
+  int dir_count = 0;
+  if ((dir = opendir(repo_path.c_str())) != NULL) {
```

Comment:
This may not be portable to other platform like Windows? Should the CMakeLists.txt be updated to disallow building with SM endpoint on Windows?

Reply:
Looks like by default the SageMaker endpoint will not be built on Windows ([ref](https://github.com/triton-inference-server/server/blob/main/build.py#L1677)), but you may add an explicit check [here](https://github.com/triton-inference-server/server/blob/main/CMakeLists.txt#L99) to disallow specifying SageMaker endpoint when `WIN32` is true